python/message.c

```diff
@@ -0,0 +1,1487 @@
+ *
+ *   # SyncSubobjs() is required to connect our existing 'sub' wrapper to the
+ *   # newly created foo.submsg data in C.
+ *   foo.MergeFrom(FooMessage(submsg={}))
+ */
+static void PyUpb_CMessage_SyncSubobjs(PyUpb_CMessage* self) {
+  PyUpb_WeakMap* subobj_map = self->unset_subobj_map;
+  upb_msg* msg = PyUpb_CMessage_GetMsg(self);
+  intptr_t iter = PYUPB_WEAKMAP_BEGIN;
+  const void* key;
+  PyObject* obj;
+
+  if (!subobj_map) return;
+
+  // The last ref to this message could disappear during iteration.
+  Py_INCREF(&self->ob_base);
+
+  while (PyUpb_WeakMap_Next(subobj_map, &key, &obj, &iter)) {
+    const upb_fielddef* f = key;
+    if (upb_fielddef_haspresence(f) && !upb_msg_has(msg, f)) continue;
+    upb_msgval msgval = upb_msg_get(msg, f);
+    PyUpb_WeakMap_DeleteIter(subobj_map, &iter);
+    if (upb_fielddef_ismap(f)) {
+      if (!msgval.map_val) continue;
+      // TODO(haberman): re-enable when maps are checked in.
+      // PyUpb_MapContainer_SwitchToSet(obj, (upb_map*)msgval.map_val);
+    } else if (upb_fielddef_isseq(f)) {
+      if (!msgval.array_val) continue;
+      // TODO(haberman): re-enable when repeated fields are checked in.
+      // PyUpb_RepeatedContainer_SwitchToSet(obj, (upb_array*)msgval.array_val);
+    } else {
+      PyUpb_CMessage* sub = (PyUpb_CMessage*)obj;
+      PyUpb_ObjCache_Add(msgval.msg_val, obj);
+      assert(self == sub->parent);
+      assert(f == PyUpb_CMessage_GetFieldDef(sub));
+      Py_DECREF((PyObject*)self);
+      sub->msg = (upb_msg*)msgval.msg_val;
+      sub->def = (uintptr_t)upb_fielddef_msgsubdef(f);
+      PyUpb_CMessage_SyncSubobjs(sub);
+    }
+  }
+
+  Py_DECREF(&self->ob_base);
+
+  // TODO(haberman): present fields need to be iterated too if they can reach
+  // a WeakMap.
+}
+
+static PyObject* PyUpb_CMessage_ToString(PyUpb_CMessage* self) {
+  if (PyUpb_CMessage_IsUnset(self)) {
+    return PyUnicode_FromStringAndSize(NULL, 0);
+  }
+  upb_msg* msg = PyUpb_CMessage_GetMsg(self);
+  const upb_msgdef* msgdef = _PyUpb_CMessage_GetMsgdef(self);
+  const upb_symtab* symtab = upb_filedef_symtab(upb_msgdef_file(msgdef));
+  char buf[1024];
+  int options = UPB_TXTENC_SKIPUNKNOWN;
+  size_t size = upb_text_encode(msg, msgdef, symtab, options, buf, sizeof(buf));
+  if (size < sizeof(buf)) {
+    return PyUnicode_FromStringAndSize(buf, size);
+  } else {
+    char* buf2 = malloc(size + 1);
+    size_t size2 =
+        upb_text_encode(msg, msgdef, symtab, options, buf2, size + 1);
+    assert(size == size2);
+    PyObject* ret = PyUnicode_FromStringAndSize(buf2, size2);
+    free(buf2);
+    return ret;
+  }
+}
+
+static PyObject* PyUpb_CMessage_RichCompare(PyObject* _self, PyObject* other,
+                                            int opid) {
+  PyUpb_CMessage* self = (PyUpb_CMessage*)_self;
+  if (opid != Py_EQ && opid != Py_NE) {
+    Py_INCREF(Py_NotImplemented);
+    return Py_NotImplemented;
+  }
+  bool ret = (opid == Py_EQ) == PyUpb_CMessage_IsEqual(self, other);
```

Comment:
I think this structure is a bit more clever than my taste.  Can we do this with two statements?
```
bool ret = IsEqual(...);
if (opid == NE) ret = !ret;
return ret;
```

https://godbolt.org/z/xfKexdE9M